Feature/#27 add new task screen

Mohamed-Boumaraf opened the pull request.

Mohamed-Boumaraf committed:
feature: create addNewTaskScreen

Mohamedragabali committed:
build:update koinAndroidxCompose
and feature: add injection AddTaskViewModel

Mohamedragabali committed:
fix:fix error in text

HendSayed25 committed:
add DatePickerModel Design

Mohamedragabali committed:
refactor: refactor code format
and add comment has handel bottom sheet

Mohamed-Boumaraf committed:
Merge branch 'feature/data-picker' of github.com:Washington-D-C-Squad/Tudee into feature/add-new-task-screen

Mohamed-Boumaraf committed:
refactor : refactoring the AddNewTaskScreen.kt

Mohamed-Boumaraf committed:
feature : handle main parts of the AddTaskScreen

Mohamedragabali committed:
feature: create bottom sheet with cancelableAction layout

Mohamed-Boumaraf committed:
feature : add category part to the screen using dummy data

Mohamedragabali committed:
refactor: refactor size of icon ,
refactor: refactor color of display button

Mohamedragabali committed:
refactor: refactor code to handel categories code

Mohamedragabali committed:
refactor: delete not important

Mohamedragabali committed:
refactor: delete not important

Mohamed-Boumaraf committed:
refactor : improve addNewTask screen design

Mohamed-Boumaraf committed:
refactor : fix dateFormatter issue

Mohamed-Boumaraf committed:
Refactor: Improve AddNewTaskScreen and TaskUiState

Mohamedragabali committed:
Merge branch 'refs/heads/develop' into feature/add-new-task-screen

# Conflicts:
#	app/src/main/java/com/washingtondcsquad/tudee/data/services/CategoriesServiceImpl.kt
#	app/src/main/java/com/washingtondcsquad/tudee/di/presentationModule.kt
#	app/src/main/java/com/washingtondcsquad/tudee/di/serviceModule.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/AppTextField.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/CancelableActionLayout.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskCard.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskPriorityCard.kt

Mohamedragabali committed:
refactor: refactor conflict in merge develop

MostafaMohamed2002 commented on app/build.gradle.kts:
```diff
@@ -63,6 +63,10 @@ dependencies {
     debugImplementation(libs.androidx.ui.test.manifest)
     ksp(libs.andoridx.room.compiler)
 
+    implementation("androidx.compose.material3:material3:1.3.2")
+    implementation("androidx.compose.material3:material3-window-size-class:1.3.2")
+    implementation("androidx.compose.material3:material3-adaptive-navigation-suite:1.4.0-alpha15")
```
why you use these dependencies ?

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/domain/entity/Task.kt:
```diff
@@ -5,10 +5,10 @@ import java.util.UUID
 data class Task(
     val id: UUID,
-    val categoryId: UUID,
+    val categoryId: Long,
     val title: String,
     val description: String,
-    val date: LocalDate,
+    val date: LocalDate? = null,
     val status: String,
-    val priority: String,
+    val priority: Priority? = null,
```
don't use null in domain entities

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/DatePickerModel.kt:
```diff
@@ -0,0 +1,80 @@
+
+            Row{
+                TextButton(onClick = onDismiss) {
+                    Text("Clear",
```
use string resource file to support localization

Mohamedragabali replied:
the owner of this task not end until now and she will resolve it in her branch to avoid conflict @HendSayed25

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddNewTaskScreen.kt:
bottom sheet it's layers above the orginal content it should be full screen

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddNewTaskScreen.kt:
```diff
@@ -0,0 +1,233 @@
+
+    val uiState by viewModel.uiState.collectAsState()
+
+    val dateAsMilliseconds = taskDate
+            .atStartOfDay(ZoneId.systemDefault())
+            .toInstant()
+            .toEpochMilli()
```
this shouold be handled in the view model

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddNewTaskScreen.kt:
```diff
@@ -0,0 +1,233 @@
+            .toInstant()
+            .toEpochMilli()
+
+    viewModel.onDateSelected(dateAsMilliseconds)
```
this is creates a side effect when this screen get recomposed this line will be executed

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddTaskViewModel.kt:
```diff
@@ -0,0 +1,109 @@
+
+class AddTaskViewModel(
+    private val tasksService: TasksService,
+    private val categoryService: CategoryLocalDataSourceImpl
```
you should inject the interface not the concerect class

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddTaskViewModel.kt:
```diff
@@ -0,0 +1,109 @@
+class AddTaskViewModel(
+    private val tasksService: TasksService,
+    private val categoryService: CategoryLocalDataSourceImpl
+) : ViewModel() {
```
you can use the base view model

MostafaMohamed2002 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddTaskViewModel.kt:
```diff
@@ -0,0 +1,109 @@
+                    description =_uiState.value.taskDescription,
+                    date = _uiState.value.taskDate,
+                    priority = _uiState.value.selectedPriority,
+                    categoryId = _uiState.value.selectedCategory!!.id,
```
this could end with a null pointer exception , be carefull

Mohamedragabali replied:
this handle from the ui logic when no selected category that will no enable button save

Mohamedragabali replied:
@MostafaMohamed2002  
thanks so much for your review :)

Ahmedsayed0895 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/DatePickerModel.kt:
```diff
@@ -0,0 +1,80 @@
+                    )
+
+                }
+                Spacer(Modifier.width(150.dp))
```
i think that you try to make your design is responsive as much as you can but be aware this hard code will break responsiveness across different screen sizes

Mohamedragabali replied:
this task of @HendSayed25  can edit that

Ahmedsayed0895 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskPriorityCard.kt:
```diff
@@ -13,46 +16,80 @@ import androidx.compose.ui.Modifier
-    Row(
+    Card(
+        colors = CardDefaults.cardColors(
+            containerColor = Color(0x00000000)
```
Great job ..  Just a small note hardcoding the transparent color could be replaced with design system Theme for better consistency and easy to update in future

Ahmedsayed0895 reviewed (commented):
overall, thank you for your hard work and your effort 👏

kareem-01 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddNewTaskScreen.kt:
no it doesn't need to be full screen, it wraps it's content and If needed would expand more to show the rest of the items, but not full screen by default

kareem-01 commented on app/src/main/java/com/washingtondcsquad/tudee/data/services/CategoriesServiceImpl.kt:
```diff
@@ -4,6 +4,7 @@ import com.washingtondcsquad.tudee.data.localSource.mapper.category.toDomain
 import com.washingtondcsquad.tudee.data.localSource.mapper.category.toEntity
 import com.washingtondcsquad.tudee.domain.entity.Category
 import com.washingtondcsquad.tudee.domain.services.CategoriesService
+import java.util.UUID
```
make sure to optimize your unused imports when commiting your work

kareem-01 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/features/sharedUiState/TaskUiState.kt:
```diff
@@ -1,10 +1,30 @@
+import com.washingtondcsquad.tudee.domain.entity.Category
+import com.washingtondcsquad.tudee.domain.entity.Priority
+import java.time.LocalDate
+
```
this is the UiState assocciated with the TaskCard, please refrain from modifying it like this and create your component's own UiState

kareem-01 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskPriorityCard.kt:
```diff
@@ -13,46 +16,80 @@ import androidx.compose.ui.Modifier
+@Composable
+private fun getBackgroundColor(
+    priority: Priority,
+    bool: Boolean
```
the name bool is not descriptive, how am i supposed what it refers to or what it's function is ?

kareem-01 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskPriorityCard.kt:
```diff
@@ -13,46 +16,80 @@ import androidx.compose.ui.Modifier
+import com.washingtondcsquad.tudee.domain.entity.Priority
+
 
 @Composable
```
this card is shared, and thus other people are using the same icon, your changes would cause other code to fail, like adding the isSelected with no default value making it required which is not the case when using the card in other screens, same thing for the onClick

kareem-01 commented:
great effort Mohamed, keep it up

saraashraf11 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskPriorityCard.kt:
```diff
@@ -13,46 +16,80 @@ import androidx.compose.ui.Modifier
+): Color {
+    return if (bool) {
+        when (priority) {
+            Priority.LOW -> AppTheme.colors.greenAccent //.colors.greenAccent
```
delete this comment

saraashraf11 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddNewTaskScreen.kt:
```diff
@@ -0,0 +1,233 @@
+        verticalArrangement = Arrangement.Bottom
+    ) {
+        ModalBottomSheet(
+            onDismissRequest = { onCancelAddTaskBottomSheet() },
```
replace the color with design system

Mohamedragabali committed:
refactor: refactor code as review

Mohamedragabali committed:
refactor: refactor not important

Mohamedragabali committed:
feature: create GetBackgroundColor composable fun

HendSayed25 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddNewTaskScreen.kt:
```diff
@@ -0,0 +1,234 @@
+                    }
+
+
+                    // priority
```
i think this comment not necessary because you use text and is clear

Mohamedragabali committed:
Merge branch 'develop' into feature/add-new-task-screen

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/washingtondcsquad/tudee/MainActivity.kt

HendSayed25 commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/screens/add_task/AddNewTaskScreen.kt:
```diff
@@ -0,0 +1,234 @@
+                            horizontalAlignment = Alignment.Start
+                        ) {
+                            Text(
+                                text = "Priority",
```
we need to move all hardcoded to string file

Mohamed-Boumaraf committed:
Refactor: Implement local data source for tasks

This commit introduces the `TaskLocalDataSource` and its implementation `TaskLocalDataSourceImpl` to handle task data persistence using Room.

Key changes include:
- Defining `DaoTask` for task-related database operations.
- Creating `TaskEntity` to represent tasks in the database.
- Implementing `TaskLocalDataSourceImpl` to interact with `DaoTask`.
- Adding `TaskMapper` for converting between domain `Task` objects and `TaskEntity`.
- Updating `TudeeDataBase` to include `DaoTask`.
- Integrating `TaskLocalDataSource` into the Koin dependency injection module.
- Modifying `TasksServiceImpl` to use the new `TaskLocalDataSource`.
- Temporarily changing `CategoriesServiceImpl` to return a hardcoded list for testing purposes.
- Updating string resources for the add new task screen.
- Modifying `MainActivity` to display `AddNewTaskScreen` directly (temporary for testing).
- Adding `TASK_TABLE_NAME` constant in `DataBaseConstants`.

karrar-abbas commented on app/src/main/java/com/washingtondcsquad/tudee/MainActivity.kt:
```diff
@@ -26,39 +27,41 @@ class MainActivity : ComponentActivity() {
+//                        }
+//                        composable(route = "category") {
+//                            // Category screen
+//                        }
```
i think the commented code should be in an individual branch if you need it in the future.

Mohamedragabali committed:
refactor: refactor code

Mohamedragabali committed:
Merge branch 'develop' into feature/add-new-task-screen

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/washingtondcsquad/tudee/MainActivity.kt
#	app/src/main/java/com/washingtondcsquad/tudee/data/localSource/TudeeDataBase.kt
#	app/src/main/java/com/washingtondcsquad/tudee/data/localSource/entities/TaskEntity.kt
#	app/src/main/java/com/washingtondcsquad/tudee/data/services/TasksServiceImpl.kt
#	app/src/main/java/com/washingtondcsquad/tudee/data/utils/DataBaseConstants.kt
#	app/src/main/java/com/washingtondcsquad/tudee/di/dataBaseModule.kt
#	app/src/main/java/com/washingtondcsquad/tudee/di/presentationModule.kt
#	app/src/main/java/com/washingtondcsquad/tudee/di/serviceModule.kt
#	app/src/main/java/com/washingtondcsquad/tudee/domain/entity/Priority.kt
#	app/src/main/java/com/washingtondcsquad/tudee/domain/entity/Task.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskCard.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/components/TaskPriorityCard.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/design/TudeeColors.kt
#	app/src/main/java/com/washingtondcsquad/tudee/presentation/features/sharedUiState/TaskUiState.kt
#	app/src/main/res/values/strings.xml
#	gradle/libs.versions.toml

Mohamedragabali committed:
refactor: resolve conflict

Mostafa-alsaygh commented on app/src/main/java/com/washingtondcsquad/tudee/data/services/FakeTasksService.kt:
```diff
@@ -5,56 +5,56 @@ import com.washingtondcsquad.tudee.domain.services.TasksService
-            status = "Pending",
-            priority = "Normal"
-        )
+//        Task(
```
don't push any commit code

Mostafa-alsaygh commented on app/src/main/java/com/washingtondcsquad/tudee/presentation/components/DatePickerModel.kt:
```diff
@@ -0,0 +1,82 @@
+//import com.washingtondcsquad.tudee.presentation.design.AppTheme
+//import com.washingtondcsquad.tudee.presentation.design.textStyle.defaultTextStyle
+//
+//@OptIn(ExperimentalMaterial3Api::class)
```
this also

Mohamedragabali committed:
refactor: refactor as reviewer

omer1998 commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/entities/TaskEntity.kt:
```diff
@@ -24,8 +26,16 @@ data class TaskEntity(
         date = date,
-        status = status,
-        priority = priority
+        status = when(status) {
```
here you will save status as string in room db why you convert it to enum

khaledeid1k replied:
why make logic here

khaledeid1k commented on app/src/main/java/com/washingtondcsquad/tudee/data/services/TaskLocalDataSource.kt:
```diff
@@ -0,0 +1,11 @@
+
+import com.washingtondcsquad.tudee.data.localSource.entities.TaskEntity
+
+interface TaskLocalDataSource {
```
we need delete this

HendSayed25 commented on app/src/main/java/com/washingtondcsquad/tudee/data/localSource/daos/DaoTask.kt:
```diff
@@ -1,12 +1,18 @@
+import java.nio.charset.CodingErrorAction.REPLACE
 
 @Dao
 interface DaoTask {
```
what difference between this interface daoTask and daoTask in local source outside daos folder ?

Mohamedragabali committed:
refactor : refactor room classes

Mohamedragabali committed:
refactor : delete not important file

HendSayed25 reviewed (approved):
Good Job 
but be careful to adding small description to the pull that will enable anyone understand what you did

Mohamedragabali committed:
Merge branch 'develop' into feature/add-new-task-screen

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/java/com/washingtondcsquad/tudee/MainActivity.kt
#	app/src/main/java/com/washingtondcsquad/tudee/di/dataBaseModule.kt
#	gradle/libs.versions.toml

Mohamedragabali committed:
Merge branch 'develop' into feature/add-new-task-screen

Mohamedragabali committed:
refactor : refactor conflict
